### PR TITLE
Anteneh.information and event features

### PR DESCRIPTION
**Pull request by AntenehAdmasu**
Error handling centralized
Data validation moved to middleware
Pagination centralized in utilities folder

### CODE REVIEW

File: Backend/controllers/EventController.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+exports.addEventLink = async(req, res) => {
+    const { id } = req.params
+    if(!req.body.linkName || !req.body.link){
+        return res.status(400).send('Link name or the link or both are missing')
+    }
+    const link = { name: req.body.linkName, link: req.body.link }
+    const event = await Event.findOne({_id: id})
+    if(!event){
+        res.status(404).send(`Event with id ${id} is not found`)
```
Comment (MohammedAlewi) [inline Backend/controllers/EventController.js:45]: return statement :)

**Review by MohammedAlewi — CHANGES_REQUESTED** [review]

**Review by MohammedAlewi — COMMENTED** [review]
looks good only some missed return statments :)

**Review by MohammedAlewi — APPROVED** [review]

**Review by Khalid-Sultan — APPROVED** [review]

File: Backend/index.js [hunk 2]
```diff
@@ -1,20 +1,31 @@
...
 
 app.use(cors());
 app.use(compression({ filter: shouldCompress }));
 app.use(bodyParser.urlencoded({ extended: false }));
 app.use(bodyParser.json());
 app.use(express.static("public"));
+app.use(router)
+
+// Centralizing error handling to avoid try catch blocks on the controllers
+app.use((err, req, res, next) => {
```
Comment (Infinidrix) [inline Backend/index.js:25]: From ExpressJS [website](https://expressjs.com/en/guide/error-handling.html#writing-error-handlers): _You define error-handling middleware last, after other app.use() and routes calls_

File: Backend/middlewares/validator.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+const _ = require('lodash')
+
+module.exports = validationSchema => {
+  return (req, res, next) => {
+    const { error } = validationSchema.validate(req.body, { abortEarly: false })
+    if (error) {
+      const response = _.map(error.details, detail => _.pick(detail, ['path', 'message']))
+      return res.status(422).send(response)
```
Comment (Infinidrix) [inline Backend/middlewares/validator.js:8]: Both this function and the centralized error handler sends the response so there will be a conflict there I think. Maybe a res.headersSent check on the universal handler would resolve the issue.

File: Backend/routes/EventRouter.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+const eventController = require('../controllers/EventController')
+const inputValidator = require('../middlewares/validator')
+const { validationSchema } = require('../models/Event')
+
+router.get('/', eventController.getAllEvents)
+router.get('/:id', eventController.getEventById)
+
+router.post('/', inputValidator(validationSchema), eventController.addEvent)
+router.put('/:id', eventController.updateEvent)
+router.put('/:id/links', eventController.addEventLink)
```
Comment (Infinidrix) [inline Backend/routes/EventRouter.js:11]: Wouldn't inputValidator be necessary for the two put methods routes

**Review by Infinidrix — COMMENTED** [review]
Left some comments but not sure if everything works like it's supposed to, would be great to have tests in the same PR as the changes.

File: Backend/index.js (hunk 2 above)
Comment (AntenehAdmasu) [reply Backend/index.js:25]: At line 30, we are using `app.use(router)` to import all routes prepared on startup/routes. The error handling middleware is invoked from `process.on(unhandledRejection)` so I don't think the position matters.

File: Backend/routes/EventRouter.js (hunk 4 above)
Comment (AntenehAdmasu) [reply Backend/routes/EventRouter.js:11]: Nice catch :) I'll update accordingly

File: Backend/middlewares/validator.js (hunk 3 above)
Comment (AntenehAdmasu) [reply Backend/middlewares/validator.js:8]: This error handler is meant to catch input errors from req.body and will be applied only when requested in the middleware, mostly for POST and PUT methods. The universal handler handles all other errors during the execution of any function.

File: Backend/index.js (hunk 2 above)
Comment (Infinidrix) [reply Backend/index.js:25]: What confused me was the `app.use(router)` on line 22. And also according to [this](https://expressjs.com/en/guide/writing-middleware.html#:~:text=Every%20time%20the%20app%20receives,first%20are%20also%20executed%20first.) I think the orderings matter like if there was an error in bodyParser then it'll catch the error, but if it is in routes it won't.

Comment (AntenehAdmasu) [reply Backend/index.js:25]: I just saw it now. We had to `app.use(router)` lines at 22 and 30 by mistake. I'll remove the later. Thanks for pointing that out :). I'm about to write the unit tests and things will settle by then I hope 👍

**Pull request closed without merging**